### PR TITLE
In-cluster config

### LINKED ISSUE #3: In-cluster config
## Description

Let's build out an in-cluster config for this so it could be loaded in the cluster via YAML files.

Might also have to think about the RBAC implications - any additional roles.

### PR DESCRIPTION
**Pull request by stefanprodan**
Fix #3 

### Description

This PR adds RBAC, Deployment and ClusterIP Service for faas-o6s.

### How Has This Been Tested?

Tested on GKE:
* create, delete and invoke functions from OpenFaaS Store using the UI
* create, update, delete and invoke functions using OpenFaaS CLI
* create, update and delete functions using kubectl

To deploy faas-o6s on an existing OpenFaaS Kubernetes setup do the following:

1. Delete all existing functions in the openfaas-fn namespace:

```bash
kubectl delete namespace openfaas-fn
kubectl create namespace openfaas-fn
```

2. Deploy faas-o6s:

```bash
# CRD
kubectl create -f artifacts/o6s-crd.yaml
# RBAC
kubectl create -f artifacts/o6s-rbac.yaml
# Service
kubectl create -f artifacts/o6s-svc.yaml
# Deployment
kubectl create -f artifacts/o6s-dep.yaml
``` 

3. Modify the gateway deployment and switch from faas-netes to faas-o6s:

```yaml
      containers:
      - name: gateway
        image: functions/gateway:0.6.16
        imagePullPolicy: Always
        env:
        - name: functions_provider_url
          value: "http://faas-o6s.openfaas:8081/"
```

### CODE REVIEW

**Comment by stefanprodan** [comment]
@alexellis can you please setup JenkinsCI

File: .travis.yml [hunk 1]
```diff
@@ -1,23 +1,27 @@
...
+  apt:
+    packages:
+      - docker-ce
+
+install:
+  - echo "Please don't go get"
 
 script:
-  - sudo docker build -t faas-o6a:latest-dev .
+  - docker build -t functions/faas-o6a:latest-dev .
```
Comment (alexellis) [inline .travis.yml]: o6a -> o6s

File: .travis.yml [hunk 2]
```diff
@@ -1,23 +1,27 @@
...
 
 after_success:
-  - if [ -z "DOCKER_USER" ]; then
+  - if [ -z "$DOCKER_USER" ]; then
       echo "PR build, skipping Docker Hub push";
     else
-      sudo docker login -u "$DOCKER_USER" -p "$DOCKER_PASS";
+      docker login -u "$DOCKER_USER" -p "$DOCKER_PASS";
+      docker push functions/faas-o6a:latest-dev;
```
Comment (alexellis) [inline .travis.yml]: o6a -> o6s

File: README.md [hunk 3]
```diff
@@ -1,12 +1,60 @@
...
+```bash
+# CRD
+kubectl create -f artifacts/o6s-crd.yaml
+# RBAC
+kubectl create -f artifacts/o6s-rbac.yaml
+# Service
+kubectl create -f artifacts/o6s-svc.yaml
+# Deployment
+kubectl create -f artifacts/o6s-dep.yaml
+```
```
[the remark this answers is not in the document]
Comment (alexellis) [reply README.md]:
```
nodeinfo.yaml
```

could go into some kind of samples folder

File: README.md [hunk 4]
```diff
@@ -1,12 +1,60 @@
...
+        imagePullPolicy: Always
+        env:
+        - name: functions_provider_url
+          value: "http://faas-o6s.openfaas:8081/"
+```
+
+Deploy a function with kubectl:
+
+```bash
+kubectl -n openfaas-fn apply -f ./artifacts/nodeinfo.yaml
```
Comment (alexellis) [inline README.md]: We used the folder name `yaml` in faas-netes.. could we use the same here?

File: artifacts/o6s-dep.yaml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      annotations:
+        prometheus.io.scrape: 'true'
+    spec:
+      serviceAccountName: faas-o6s
+      containers:
+      - name: faas-o6s
+        image: stefanprodan/faas-o6s:latest-dev
+        imagePullPolicy: Always
+        command:
+          - ./faas-o6s
```
Comment (alexellis) [inline artifacts/o6s-dep.yaml]: We generally don't configure the other OpenFaaS service via `command` but through environmental variables.

File: pkg/controller/deployment.go [hunk 6]
```diff
@@ -112,9 +113,9 @@ func makeLivenessProbe() *corev1.Probe {
 		},
 		InitialDelaySeconds: 3,
 		TimeoutSeconds:      1,
-		PeriodSeconds:       10,
+		PeriodSeconds:       5,
 		SuccessThreshold:    1,
-		FailureThreshold:    3,
+		FailureThreshold:    2,
```
Comment (alexellis) [inline pkg/controller/deployment.go]: Does PeriodSeconds and FailureThreshold match the current values in faas-netes?

File: artifacts/o6s-dep.yaml (hunk 5 above)
Comment (stefanprodan) [reply artifacts/o6s-dep.yaml]: there is no way around this, glog only works with cmd flags

File: pkg/controller/deployment.go (hunk 6 above)
Comment (stefanprodan) [reply pkg/controller/deployment.go]: This ensures max 10s of downtime, the faas-netes settings should change also.

File: README.md (hunk 4 above)
Comment (stefanprodan) [reply README.md]: That naming is part of the controller template, we could change it when we port it to faas-netes.

File: artifacts/o6s-dep.yaml (hunk 5 above)
Comment (alexellis) [reply artifacts/o6s-dep.yaml]: even using our own binary? That's surprising.. maybe we can PR it? :laughing:

File: pkg/controller/deployment.go (hunk 6 above)
Comment (alexellis) [reply pkg/controller/deployment.go]: Please raise a PR to faas-netes too then.

**Pull request merged**